core/src/main/java/org/apache/calcite/rel/rules/AggregateCaseToFilterRule.java

```diff
@@ -227,8 +227,11 @@ protected AggregateCaseToFilterRule(RelBuilderFactory relBuilderFactory,
           RelCollations.EMPTY, aggregateCall.getType(),
           aggregateCall.getName());
     } else if (kind == SqlKind.SUM // Case B
-        && isIntLiteral(arg1) && RexLiteral.intValue(arg1) == 1
-        && isIntLiteral(arg2) && RexLiteral.intValue(arg2) == 0) {
+        && isIntLiteral(arg1)
+        && RexLiteral.intValue(arg1) == 1
+        && isIntLiteral(arg2)
+        && !RexLiteral.isNullLiteral(arg2)
```

Comment:
Is it even possible for `RexLiteral.isNullLiteral(arg2)` to return true after `isIntLiteral(arg2)` has returned true?

If so, make `isIntLiteral(arg2)` only return true if the value is not null.

Reply:
The origin NPE is `isIntLiteral(arg2)` return true but arg2's value is null.

`isIntLiteral` doesn't consider the value:
```
return rexNode instanceof RexLiteral && SqlTypeName.INT_TYPES.contains(rexNode.getType().getSqlTypeName());
```

Reply:
Then there might be a rounding issue. Currently `isIntLiteral` returns true for `TINYINT, SMALLINT, INTEGER, BIGINT`. Apparently, `intValue` would return the wrong result in case `BigDecimal` contains values like `123456789012345678901234567890` that will be truncated. It should be `BigDecimal#intValueExact()` rather than `intValue()`

Reply:
if `isIntLiteral(decimal)` returns false, why do we still use `intValue(decimal)`?

Reply:
The previous code, with two conditions per line, made the logic really clear:
```
    && isIntLiteral(arg1) && RexLiteral.intValue(arg1) == 1
    && isIntLiteral(arg2) && RexLiteral.intValue(arg2) == 0
```

We just weren't handling some details about null values and - as @vlsi points out - values outside the int range. Let's go back to the old structure.

Make the necessary changes to `isIntLiteral` to deal with the corner cases. It's private so you can do what you need to.

Reply:
But is "null" a reasonable int literal? @vlsi 's case, "isIntLiteral(decimal)" just returns false. My case is "null" is an int literal but you when you get its value, it is null.

Reply:
@julianhyde If we think null is not a reasonable int literal, I can add this cond to "isIntLiteral", but the decimal case is not the same, if isIntLiteral(decimal) returns false, we already know it is not an int, why do we still use intValue(decimal)?